core/metacling/src/TCling.cxx

```diff
@@ -6069,61 +6069,63 @@ Int_t TCling::ShallowAutoLoadImpl(const char *cls)
-               }
-            }
-         } else {
-            auto classinfo = gInterpreter->ClassInfo_Factory(cls);
-            if (classinfo && gInterpreter->ClassInfo_IsValid(classinfo)
-                && !(gInterpreter->ClassInfo_Property(classinfo) & kIsEnum))
-            {
-               DataMemberInfo_t *memberinfo = gInterpreter->DataMemberInfo_Factory(classinfo, TDictionary::EMemberSelection::kNoUsingDecls);
-               while (gInterpreter->DataMemberInfo_Next(memberinfo)) {
-                  auto membertypename = TClassEdit::GetLong64_Name(gInterpreter->TypeName(gInterpreter->DataMemberInfo_TypeTrueName(memberinfo)));
-                  if (!(gInterpreter->DataMemberInfo_TypeProperty(memberinfo) & ::kIsFundamental)
-                      && !TClassTable::GetDictNorm(membertypename.c_str()))
-                  {
-                     // Failure to load a dictionary is not (quite) a failure load
-                     // the top-level library.   See detailed comment in the TProtoClass
-                     // branch (above).
-                     (void)DeepAutoLoadImpl(membertypename.c_str());
-                  }
-               }
-               gInterpreter->DataMemberInfo_Delete(memberinfo);
-            }
-            gInterpreter->ClassInfo_Delete(classinfo);
+// \note `cls` is expected to be already normalized!
+// \returns 1 on success.
+Int_t TCling::DeepAutoLoadImpl(const char *cls, std::unordered_set<std::string> &visited,
+                               bool nameIsNormalized)
+{
+   // Try to insert; if insertion failed because the entry existed, DeepAutoLoadImpl()
+   // has previously (within the same call to `AutoLoad()`) tried to load this class
+   // and we are done, whether success or not, as it won't work better now than before,
+   // because there is no additional information now compared to before.
+   if (!visited.insert(std::string(cls)).second)
+      return 1;
+
+   if (ShallowAutoLoadImpl(cls) == 0) {
+      // If ShallowAutoLoadImpl() has an error, we have an error.
+      return 0;
+   }
+
+   // Now look through the TProtoClass to load the required library/dictionary
+   if (TProtoClass *proto = nameIsNormalized ? TClassTable::GetProtoNorm(cls) : TClassTable::GetProto(cls)) {
+      for (auto element : proto->GetData()) {
+         if (element->IsBasic())
+            continue;
+         const char *subtypename = element->GetTypeName();
+         if (!TClassTable::GetDictNorm(subtypename)) {
+            // Failure to load a dictionary is not (quite) a failure load
+            // the top-level library.  If we return false here, then
+            // we would end up in a situation where the library and thus
+            // the dictionary is loaded for "cls" but the TClass is
+            // not created and/or marked as unavailable (in case where
+            // AutoLoad is called from TClass::GetClass).
+            DeepAutoLoadImpl(subtypename, visited, true /*normalized*/);
```

Comment:
The cast to `(void)` end up here because some compiler complained that the result was not used.

Reply:
OK but we don't seem to see those anymore?